presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/RemoveUnsupportedDynamicFilters.java

```diff
@@ -181,12 +187,41 @@ private Expression removeDynamicFilters(Expression expression, Set<String> allow
 
         private Expression removeAllDynamicFilters(Expression expression)
         {
-            DynamicFilters.ExtractResult extractResult = extractDynamicFilters(expression);
+            Expression rewrittenExpression = removeNestedDynamicFilters(expression);
+            DynamicFilters.ExtractResult extractResult = extractDynamicFilters(rewrittenExpression);
             if (extractResult.getDynamicConjuncts().isEmpty()) {
-                return expression;
+                return rewrittenExpression;
             }
             return combineConjuncts(metadata, extractResult.getStaticConjuncts());
         }
+
+        private Expression removeNestedDynamicFilters(Expression expression)
+        {
+            return ExpressionTreeRewriter.rewriteWith(new RemoveDynamicFilters(), expression);
+        }
+    }
+
+    private static class RemoveDynamicFilters
+            extends io.prestosql.sql.tree.ExpressionRewriter<Void>
+    {
+        @Override
+        public Expression rewriteLogicalBinaryExpression(LogicalBinaryExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            LogicalBinaryExpression rewrittenNode = treeRewriter.defaultRewrite(node, context);
+
+            ImmutableList.Builder<Expression> expressionBuilder = ImmutableList.builder();
+            if (!isDynamicFilter(rewrittenNode.getLeft())) {
+                expressionBuilder.add(rewrittenNode.getLeft());
+            }
+            if (!isDynamicFilter(rewrittenNode.getRight())) {
+                expressionBuilder.add(rewrittenNode.getRight());
+            }
+            Expression modified = binaryExpression(node.getOperator(), expressionBuilder.build());
+            if (modified.equals(node)) {
```

Comment:
compare expressions using identity `==`

Reply:
it becomes equivalent to just `return binaryExpression(node.getOperator(), expressionBuilder.build());`

Reply:
`binaryExpression` will create new expression instance

Reply:
Would below be the right way of short-circuiting here ?
```
List<Expression> staticPredicates = expressionBuilder.build();
if ((node == rewrittenNode) && staticPredicates.size() == 2) {
     return node;
}
return binaryExpression(node.getOperator(), staticPredicates);
```